### PR TITLE
Add Main Paymaster Example Components

### PR DESCRIPTION
**Pull request by cpcramer**
**What changed? Why?**
- Add paymaster proxy route. 
- Add paymaster documentation guide.
- Add frontend styling. 
- Add willSponsor helper function for client side verification. 
- General refactoring and cleanup. 

Screenshot 2024-05-21 at 9 12 37 AM

**Notes to reviewers**

**How has it been tested?**

### CODE REVIEW

File: template/web/app/paymaster-bundler/utils/willSponsor.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+/**
+ * willSponsor function checks if a user operation should be sponsored based on several criteria.
+ *
+ * @param {Object} params - The parameters for the function.
+ * @param {number} params.chainId - The chain ID to check.
+ * @param {string} params.entrypoint - The entrypoint address to check.
+ * @param {UserOperation<"v0.6">} params.userOp - The user operation object containing the call data.
+ * @returns {Promise<boolean>} - Returns true if the user operation meets all criteria and should be sponsored, otherwise false.
+ */
+// TODO: Split this into smaller chunks and try to have tests for this.
```
Comment (cpcramer) [inline template/web/app/paymaster-bundler/utils/willSponsor.ts:24]: Will split up and add tests.

File: template/web/app/api/paymasterProxy/route.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+};
+
+/**
+ * POST function handles incoming POST requests for sponsoring operations.
+ *
+ * @param {Request} r - The incoming request object.
+ * @returns {Promise<Response>} - The response object containing the result or an error message.
+ */
+// TODO: ADD tests for all of this file.
+export async function POST(req: NextRequest): Promise<NextResponse> {
```
Comment (cpcramer) [inline template/web/app/api/paymasterProxy/route.ts:18]: Will add tests.

File: template/web/src/components/SmartWallets/PaymasterClient.ts [hunk 3]
```diff
@@ -1,7 +1,16 @@
...
+import { createClient, createPublicClient, http } from 'viem';
 import { baseSepolia } from 'viem/chains';
 
+const paymasterService = process.env.NEXT_PUBLIC_PAYMASTER_URL ?? '';
+
 export const client = createPublicClient({
   chain: baseSepolia,
   transport: http(),
 });
+
```
Comment (Zizzamia) [inline template/web/src/components/SmartWallets/PaymasterClient.ts:12]: Why do we have two clients?

Maybe is fair to explain that.

File: template/web/src/components/SmartWallets/PaymasterClient.ts [hunk 4]
```diff
@@ -1,7 +1,16 @@
-import { createPublicClient, http } from 'viem';
+import { ENTRYPOINT_ADDRESS_V06 } from 'permissionless';
+import { paymasterActionsEip7677 } from 'permissionless/experimental';
+import { createClient, createPublicClient, http } from 'viem';
 import { baseSepolia } from 'viem/chains';
 
+const paymasterService = process.env.NEXT_PUBLIC_PAYMASTER_URL ?? '';
```
Comment (Zizzamia) [inline template/web/src/components/SmartWallets/PaymasterClient.ts:6]: Was this in the config?

File: template/web/src/components/SmartWallets/PaymasterClient.ts (hunk 3 above)
Comment (Zizzamia) [reply template/web/src/components/SmartWallets/PaymasterClient.ts:12]: Also should this client shared across all experiences in BOAT?

File: template/web/package.json [hunk 5]
```diff
@@ -21,7 +21,6 @@
     "@coinbase/onchainkit": "0.16.10",
     "@coinbase/wallet-sdk": "4.0.0",
     "@farcaster/hub-nodejs": "^0.11.9",
-    "@privy-io/react-auth": "^1.56.2",
```
Comment (Zizzamia) [inline template/web/package.json:24]: ok dope

File: template/web/jest.config.js [hunk 6]
```diff
@@ -8,7 +8,11 @@ const customJestConfig = {
   moduleNameMapper: {
     'rehype-pretty-code': '<rootDir>/node_modules/rehype-pretty-code',
   },
-  roots: ['<rootDir>/src'],
+  roots: [
+    '<rootDir>/src',
+    '<rootDir>/app/api/paymasterProxy',
+    '<rootDir>/app/paymaster-bundler/utils/',
```
[the remark this answers is not in the document]
Comment (Zizzamia) [reply template/web/jest.config.js:14]: Interesting..... if that's the case also open a public issue on Github so we don't forget to fix it.

File: template/web/app/paymaster-bundler/utils/willSponsor.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+/**
+ * willSponsor function checks if a user operation should be sponsored based on several criteria.
+ *
+ */
+export async function willSponsor({
+  chainId,
+  entrypoint,
+  userOp,
+}: {
```
Comment (Zizzamia) [inline template/web/app/paymaster-bundler/utils/willSponsor.ts:64]: Types are easier to read when are more explicit as their own variable.

like

type dude = { ... }

and than you use `dude`

File: template/web/app/paymaster-bundler/utils/willSponsor.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  chainId: number;
+  entrypoint: string;
+  userOp: UserOperation<'v0.6'>;
+}): Promise<boolean> {
+  if (!isValidChainAndEntryPoint(chainId, entrypoint)) return false;
+
+  try {
+    if (!isValidCallData(userOp)) return false;
+
```
Comment (Zizzamia) [inline template/web/app/paymaster-bundler/utils/willSponsor.ts:73]: Why this empty space?

File: template/web/app/paymaster-bundler/utils/willSponsor.ts [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+export async function willSponsor({
+  chainId,
+  entrypoint,
+  userOp,
+}: {
+  chainId: number;
+  entrypoint: string;
+  userOp: UserOperation<'v0.6'>;
+}): Promise<boolean> {
+  if (!isValidChainAndEntryPoint(chainId, entrypoint)) return false;
```
Comment (Zizzamia) [inline template/web/app/paymaster-bundler/utils/willSponsor.ts:69]: Always use proper JS with `{ }`, never cut corner with one-liner.
one-liner JS are a sleppery slope for code that is harder to read, and easier to break.

File: template/web/app/paymaster-bundler/utils/willSponsor.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  }[];
+  if (calls.length > 2) return false;
+
+  let callToCheckIndex = 0;
+  if (calls.length > 1) {
+    if (calls[0].target.toLowerCase() !== MAGIC_SPEND_ADDRESS.toLowerCase()) return false;
+    callToCheckIndex = 1;
+  }
+
+  if (calls[callToCheckIndex].target.toLowerCase() !== PAYMASTER_NFT_ADDRESS.toLowerCase())
```
Comment (Zizzamia) [inline template/web/app/paymaster-bundler/utils/willSponsor.ts:44]: Please write proper conditional. this code is hard to read.

File: template/web/app/paymaster-bundler/utils/willSponsor.ts [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+function isValidCallData(userOp: UserOperation<'v0.6'>): boolean {
+  const calldata = decodeFunctionData({
+    abi: CoinbaseSmartWalletABI,
+    data: userOp.callData,
+  });
+
+  if (calldata.functionName !== 'executeBatch') return false;
+  if (!calldata.args || calldata.args.length == 0) return false;
+
+  const calls = calldata.args[0] as {
```
Comment (Zizzamia) [inline template/web/app/paymaster-bundler/utils/willSponsor.ts:31]: Each type should have his own variable.

File: template/web/app/paymaster-bundler/utils/willSponsor.ts [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+ */
+function isValidChainAndEntryPoint(chainId: number, entrypoint: string): boolean {
+  if (chainId !== baseSepolia.id) return false;
+  if (entrypoint.toLowerCase() !== ENTRYPOINT_ADDRESS_V06.toLowerCase()) return false;
+  return true;
+}
+
+/**
+ * Checks if the call data in the user operation is valid and meets sponsorship criteria.
+ *
```
Comment (Zizzamia) [inline template/web/app/paymaster-bundler/utils/willSponsor.ts:20]: Why an empty string?

File: template/web/app/paymaster-bundler/_components/PaymasterBundlerDemo.tsx [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+  const contract = usePaymasterBundlerContract();
+
+  if (contract.status !== 'ready') {
+    console.error('Contract is not ready');
+    return null;
+  }
+
+  const handleMint = () => {
+    console.log('Default url: ', defaultUrl);
+    console.log('account: ', account);
```
Comment (Zizzamia) [inline template/web/app/paymaster-bundler/_components/PaymasterBundlerDemo.tsx:30]: All this console.log are all over the place. If want to use them, we should have a more clean format.

File: template/web/app/paymaster-bundler/_components/PaymasterBundlerDemo.tsx [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+import { useWriteContracts } from 'wagmi/experimental';
+import isLocal from '../../../src/utils/isLocal';
+import { usePaymasterBundlerContract } from '../_contracts/usePaymasterBundlerContract';
+import { CallStatus } from './CallStatus';
+
+// Use the local API URL to target the Paymaster directly without a proxy
+// if running on localhost, otherwise use the Paymaster Proxy.
+const isLocalEnv = isLocal();
+const defaultUrl = isLocalEnv
+  ? `https://api.developer.coinbase.com/rpc/v1/base-sepolia/z7inYI-NRNAOF9kgaW4Suf-30N6DuMra`
```
Comment (Zizzamia) [inline template/web/app/paymaster-bundler/_components/PaymasterBundlerDemo.tsx:12]: those should probably be constant.

File: template/web/app/api/paymasterProxy/route.ts [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+  params: [UserOperation<'v0.6'>, string, string];
+};
+
+/**
+ * POST function handles incoming POST requests for sponsoring operations.
+ *
+ * @param {Request} r - The incoming request object.
+ * @returns {Promise<Response>} - The response object containing the result or an error message.
+ */
+// TODO: ADD tests for all of this file.
```
Comment (Zizzamia) [inline template/web/app/api/paymasterProxy/route.ts:17]: What about here?

**Review by esmaili76000 — CHANGES_REQUESTED** [review]
[]()~~~~~~~~~~~~~~~~

**Review by esmaili76000 — APPROVED** [review]

File: template/web/src/components/SmartWallets/PaymasterClient.ts (hunk 3 above)
Comment (cpcramer) [reply template/web/src/components/SmartWallets/PaymasterClient.ts:12]: Ah thanks that is an unused client - removed

File: template/web/src/components/SmartWallets/PaymasterClient.ts (hunk 4 above)
Comment (cpcramer) [reply template/web/src/components/SmartWallets/PaymasterClient.ts:6]: This was originally in the `constants.ts` file. I figured that it wasn't necessary to assign a const to the env variable when I can just access the env whenever I needed it. 

It looked like:
`const PAYMASTER_SERVICE_URL = process.env.NEXT_PUBLIC_PAYMASTER_URL`

File: template/web/app/paymaster-bundler/utils/willSponsor.ts (hunk 8 above)
Comment (cpcramer) [reply template/web/app/paymaster-bundler/utils/willSponsor.ts:73]: fixed!

File: template/web/app/paymaster-bundler/utils/willSponsor.ts (hunk 9 above)
Comment (cpcramer) [reply template/web/app/paymaster-bundler/utils/willSponsor.ts:69]: > Simple is better than complex. 
> Readability counts.
> 

Fixed!

File: template/web/app/paymaster-bundler/utils/willSponsor.ts (hunk 10 above)
Comment (cpcramer) [reply template/web/app/paymaster-bundler/utils/willSponsor.ts:44]: > Simple is better than complex. 
> Readability counts.
> 

Fixed!

File: template/web/app/paymaster-bundler/utils/willSponsor.ts (hunk 11 above)
Comment (cpcramer) [reply template/web/app/paymaster-bundler/utils/willSponsor.ts:31]: Updated to:

```
type CallDataTargetAddress = Address;
type CallDataValue = bigint;
type CallData = Hex;

  const calls = calldata.args[0] as {
    target: CallDataTargetAddress;
    value: CallDataValue;
    data: CallData;
  }
```

File: template/web/app/paymaster-bundler/utils/willSponsor.ts [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+ */
+function isValidCallData(userOp: UserOperation<'v0.6'>): boolean {
```
[the remark this answers is not in the document]
Comment (cpcramer) [reply template/web/app/paymaster-bundler/utils/willSponsor.ts:22]: I know this looks weird but it's the version that UserOperation extends...there is a type error if this is omitted.

[the remark this answers is not in the document]
Comment (cpcramer) [reply template/web/app/paymaster-bundler/utils/willSponsor.ts:22]: From the documentation - `export type EntryPointVersion = "v0.6" | "v0.7"`

File: template/web/app/paymaster-bundler/utils/willSponsor.ts (hunk 12 above)
Comment (cpcramer) [reply template/web/app/paymaster-bundler/utils/willSponsor.ts:20]: I don't follow, should I add more to the function comment?

File: template/web/app/paymaster-bundler/_components/PaymasterBundlerDemo.tsx (hunk 13 above)
Comment (cpcramer) [reply template/web/app/paymaster-bundler/_components/PaymasterBundlerDemo.tsx:30]: Yeah sorry these were more for me to help debug. Removed!

File: template/web/app/paymaster-bundler/_components/PaymasterBundlerDemo.tsx (hunk 14 above)
Comment (cpcramer) [reply template/web/app/paymaster-bundler/_components/PaymasterBundlerDemo.tsx:12]: Added `PAYMASTER_URL` constant

File: template/web/app/api/paymasterProxy/route.ts (hunk 15 above)
Comment (cpcramer) [reply template/web/app/api/paymasterProxy/route.ts:17]: Whoops . These are done. Removed!

File: template/web/jest.config.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (cpcramer) [reply template/web/jest.config.js:14]: Tracking here https://github.com/coinbase/build-onchain-apps/issues/496

File: template/web/src/components/SmartWallets/PaymasterClient.ts (hunk 3 above)
Comment (cpcramer) [reply template/web/src/components/SmartWallets/PaymasterClient.ts:13]: This client is used in `willSponsor.test.ts`. I tried to define it directly in the test file but it doesn't work...getting type errors.

Comment (cpcramer) [reply template/web/src/components/SmartWallets/PaymasterClient.ts:13]: > Also should this client shared across all experiences in BOAT?

I'm not sure. Can refactor if we see that this client is needed elsewhere?

File: template/web/app/paymaster-bundler/utils/willSponsor.ts (hunk 12 above)
Comment (cpcramer) [reply template/web/app/paymaster-bundler/utils/willSponsor.ts:38]: Wait - Updated to remove the empty * line

Comment (esmaili76000) [reply template/web/app/paymaster-bundler/utils/willSponsor.ts]: - - - [ ] > @@@@[]()

File: template/web/src/components/SmartWallets/PaymasterClient.ts [hunk 17]
```diff
@@ -1,7 +1,17 @@
-import { createPublicClient, http } from 'viem';
+import { ENTRYPOINT_ADDRESS_V06 } from 'permissionless';
+import { paymasterActionsEip7677 } from 'permissionless/experimental';
+import { createClient, http } from 'viem';
+import { createPublicClient } from 'viem';
```
Comment (Zizzamia) [inline template/web/src/components/SmartWallets/PaymasterClient.ts:4]: why two lines to import from `viem`, just do one line.

File: template/web/src/components/SmartWallets/PaymasterClient.ts (hunk 3 above)
Comment (Zizzamia) [reply template/web/src/components/SmartWallets/PaymasterClient.ts:13]: let's keep it for future refactor, either way seems something that can be reused.

File: template/web/src/components/SmartWallets/PaymasterClient.ts (hunk 17 above)
Comment (cpcramer) [reply template/web/src/components/SmartWallets/PaymasterClient.ts]: Whoops, fixed!

Comment (esmaili76000) [reply template/web/src/components/SmartWallets/PaymasterClient.ts]: > - [ ] @

**Pull request closed without merging**